File: java/131.palindrome-partitioning.java
/*
 * @lc app=leetcode id=131 lang=java
 *
 * [131] Palindrome Partitioning
 */

// @lc code=start
class Solution {

    boolean[][] dp;

    List<List<String>> res = new LinkedList<>();

    public List<List<String>> partition(String s) {
        if (s != null) {
            initPalindromeIndicator(s);
            partitioning(new Stack<>(), 0, s);
        }
        return res;
    }

    private void partitioning(Stack<String> stack, int start, String s) {
        if (start == s.length()) {
            res.add(new ArrayList<>(stack));
            return;
        }
        for (int i = start; i < s.length(); ++i) {
            if (isPalindrome(start, i)) {
                stack.push(s.substring(start, i+1));
                partitioning(stack, i+1, s);
                stack.pop();
            }
        }
    }

    private void initPalindromeIndicator(String s) {
        if (s.length() == 0) return;
        dp = new boolean[s.length()][s.length()];
        for (int i = 0; i < s.length(); ++i) {
            for (int j = 0; j <= i; ++j) {
                if (s.charAt(i) == s.charAt(j) && (i - j <= 2 || dp[j+1][i-1])) dp[j][i] = true;
            }
        }
    }

    private boolean isPalindrome(int begin, int end) {
        return dp[begin][end];
    }
}
// @lc code=end

File: go/131.palindrome-partitioning.go
/*
 * @lc app=leetcode id=131 lang=golang
 *
 * [131] Palindrome Partitioning
 */

// @lc code=start
func partition(s string) [][]string {
	res := [][]string{}
	
	var backtrack func(int, []string)
	backtrack = func(idx int, pt []string) {
		if idx == len(s) {
			tmp := make([]string, len(pt))
			copy(tmp, pt)
			res = append(res, tmp)
			return
		}
		for i := idx; i < len(s); i++ {
			if isPalindrome(s[idx:i+1]) {
				pt = append(pt, s[idx:i+1])
				backtrack(i+1, pt)
				pt = pt[:len(pt)-1]
			}
		}
	}

	backtrack(0, []string{})
	return res
}

func isPalindrome(s string) bool {
	if len(s) == 1 {
		return true
	}
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}
// @lc code=end

